### ISSUE TITLE
Major bug discovered in golang or perhaps bug in our code.

### ISSUE THREAD

**Issue by harshavardhana** [minio/mc#1260]
Sample recursive output from S3. 

```
[2015-09-05 01:39:18 PDT]   215B ferenginar/Go/blog/content/constants/syntax.go
[2015-09-05 01:39:19 PDT] 9.3KiB ferenginar/Go/blog/content/context.article
[2015-09-05 01:39:18 PDT] 2.6KiB ferenginar/Go/blog/content/context/google/google.go
[2015-09-05 01:39:19 PDT] 1.4KiB ferenginar/Go/blog/content/context/gorilla/gorilla.go
[2015-09-05 01:39:19 PDT] 1.7KiB ferenginar/Go/blog/content/context/interface.go
[2015-09-05 01:39:19 PDT] 2.7KiB ferenginar/Go/blog/content/context/server/server.go
[2015-09-05 01:39:19 PDT]   557B ferenginar/Go/blog/content/context/tomb/tomb.go
[2015-09-05 01:39:19 PDT] 1.4KiB ferenginar/Go/blog/content/context/userip/userip.go
[2015-09-05 01:39:20 PDT]  12KiB ferenginar/Go/blog/content/cover.article
[2015-09-05 01:39:19 PDT]  59KiB ferenginar/Go/blog/content/cover/count.png
[2015-09-05 01:39:19 PDT]   786B ferenginar/Go/blog/content/cover/pkg.cover
[2015-09-05 01:39:20 PDT]   230B ferenginar/Go/blog/content/cover/pkg.go
[2015-09-05 01:39:20 PDT]   318B ferenginar/Go/blog/content/cover/pkg_test.go
[2015-09-05 01:39:20 PDT]  42KiB ferenginar/Go/blog/content/cover/set.png
[2015-09-05 01:39:20 PDT] 3.1KiB ferenginar/Go/blog/content/debugging-go-code-status-report.article
[2015-09-05 01:39:20 PDT]   760B ferenginar/Go/blog/content/debugging-go-programs-with-gnu-debugger.article

```

Sample recursive output from fs

```
[2015-11-11 01:39:28 PST]  20KiB test/ferenginar/Go/blog/content/constants.article
[2015-11-11 01:20:14 PST] 2.6KiB test/ferenginar/Go/blog/content/context/google/google.go
[2015-11-11 01:20:14 PST] 1.4KiB test/ferenginar/Go/blog/content/context/gorilla/gorilla.go
[2015-11-11 01:20:14 PST] 1.7KiB test/ferenginar/Go/blog/content/context/interface.go
[2015-11-11 01:20:14 PST] 2.7KiB test/ferenginar/Go/blog/content/context/server/server.go
[2015-11-11 01:20:15 PST]   557B test/ferenginar/Go/blog/content/context/tomb/tomb.go
[2015-11-11 01:20:15 PST] 1.4KiB test/ferenginar/Go/blog/content/context/userip/userip.go
[2015-11-11 01:39:27 PST] 9.3KiB test/ferenginar/Go/blog/content/context.article
[2015-11-11 01:20:15 PST]  59KiB test/ferenginar/Go/blog/content/cover/count.png
[2015-11-11 01:20:15 PST]   786B test/ferenginar/Go/blog/content/cover/pkg.cover
[2015-11-11 01:20:15 PST]   230B test/ferenginar/Go/blog/content/cover/pkg.go
[2015-11-11 01:20:15 PST]   318B test/ferenginar/Go/blog/content/cover/pkg_test.go
[2015-11-11 01:20:15 PST]  42KiB test/ferenginar/Go/blog/content/cover/set.png
[2015-11-11 01:39:28 PST]  12KiB test/ferenginar/Go/blog/content/cover.article
[2015-11-11 01:20:15 PST] 3.1KiB test/ferenginar/Go/blog/content/debugging-go-code-status-report.article
[2015-11-11 01:20:15 PST]   760B test/ferenginar/Go/blog/content/debugging-go-programs-with-gnu-debugger.article
```

If you can clearly observe.  Both are providing sorting but the basis of sort is different ie. 

on S3 cover.article precedes cover/count.png

```
[2015-09-05 01:39:20 PDT]  12KiB ferenginar/Go/blog/content/cover.article
[2015-09-05 01:39:19 PDT]  59KiB ferenginar/Go/blog/content/cover/count.png
``` 

on FS cover.article comes last. 

```
[2015-11-11 01:20:15 PST]  59KiB test/ferenginar/Go/blog/content/cover/count.png
[2015-11-11 01:20:15 PST]   786B test/ferenginar/Go/blog/content/cover/pkg.cover
[2015-11-11 01:20:15 PST]   230B test/ferenginar/Go/blog/content/cover/pkg.go
[2015-11-11 01:20:15 PST]   318B test/ferenginar/Go/blog/content/cover/pkg_test.go
[2015-11-11 01:20:15 PST]  42KiB test/ferenginar/Go/blog/content/cover/set.png
[2015-11-11 01:39:28 PST]  12KiB test/ferenginar/Go/blog/content/cover.article
```

**Comment by harshavardhana** [comment]
Reducing the solution space. 

```
$ mc ls test/ferenginar/Go/blog/content/cover...
[2015-11-11 01:20:15 PST]  59KiB test/ferenginar/Go/blog/content/cover/count.png
[2015-11-11 01:20:15 PST]   786B test/ferenginar/Go/blog/content/cover/pkg.cover
[2015-11-11 01:20:15 PST]   230B test/ferenginar/Go/blog/content/cover/pkg.go
[2015-11-11 01:20:15 PST]   318B test/ferenginar/Go/blog/content/cover/pkg_test.go
[2015-11-11 01:20:15 PST]  42KiB test/ferenginar/Go/blog/content/cover/set.png
[2015-11-11 01:39:28 PST]  12KiB test/ferenginar/Go/blog/content/cover.article
``` 

```
$ mc ls s3/ferenginar/Go/blog/content/cover...
[2015-09-05 01:39:20 PDT]  12KiB ferenginar/Go/blog/content/cover.article
[2015-09-05 01:39:19 PDT]  59KiB ferenginar/Go/blog/content/cover/count.png
[2015-09-05 01:39:19 PDT]   786B ferenginar/Go/blog/content/cover/pkg.cover
[2015-09-05 01:39:20 PDT]   230B ferenginar/Go/blog/content/cover/pkg.go
[2015-09-05 01:39:20 PDT]   318B ferenginar/Go/blog/content/cover/pkg_test.go
[2015-09-05 01:39:20 PDT]  42KiB ferenginar/Go/blog/content/cover/set.png
```

**Comment by harshavardhana** [comment]
Now we can really observe what is going on, the difference is evident - both are lexically sorted as they claim. But one of them should be correct.

**Comment by harshavardhana** [comment]
Removing even "test/" directory and making sure to be equal to s3 same result for "fs"

```
$ mc ls ferenginar/Go/blog/content/cover...
[2015-11-11 01:55:37 PST]  59KiB ferenginar/Go/blog/content/cover/count.png
[2015-11-11 01:55:37 PST]   786B ferenginar/Go/blog/content/cover/pkg.cover
[2015-11-11 01:55:37 PST]   230B ferenginar/Go/blog/content/cover/pkg.go
[2015-11-11 01:55:37 PST]   318B ferenginar/Go/blog/content/cover/pkg_test.go
[2015-11-11 01:55:38 PST]  42KiB ferenginar/Go/blog/content/cover/set.png
[2015-11-11 01:55:37 PST]  12KiB ferenginar/Go/blog/content/cover.article
```

**Comment by harshavardhana** [comment]
```
[2015-11-11 03:04:45 PST] 9.3KiB newbucket/Go/blog/content/context.article
[2015-11-11 03:04:45 PST]  59KiB newbucket/Go/blog/content/cover/count.png
[2015-11-11 03:04:45 PST]   786B newbucket/Go/blog/content/cover/pkg.cover
[2015-11-11 03:04:45 PST]   230B newbucket/Go/blog/content/cover/pkg.go
[2015-11-11 03:04:45 PST]   318B newbucket/Go/blog/content/cover/pkg_test.go
[2015-11-11 03:04:45 PST]  42KiB newbucket/Go/blog/content/cover/set.png
[2015-11-11 03:04:45 PST]  12KiB newbucket/Go/blog/content/cover.article
```

minio server on the other hand implements like this.

**Comment by harshavardhana** [comment]
A simplest program to reproduce the behavior. 

```
$ go run walk.go
test
test/new
test/new/old.txt
test/new.txt
```

And simple listobjects.go 

```
$ go run listobjects.go
2015/11/11 11:44:08 test.txt
2015/11/11 11:44:08 test/new.txt
2015/11/11 11:44:08 test/new/old.txt
```

**Comment by harshavardhana** [comment]
```
package main

import (
	"fmt"
	"os"
	"path/filepath"
)

func main() {
	walkFn := func(fp string, info os.FileInfo, err error) error {
		fmt.Println(fp)
		return nil
	}
	filepath.Walk("test", walkFn)
}
```

**Comment by harshavardhana** [comment]
```
package main

import (
	"log"

	"github.com/minio/attic/minio-go-legacy"
)

func main() {
	config := minio.Config{
		AccessKeyID:     "**********",
		SecretAccessKey: "**********",
		Endpoint:        "https://s3.amazonaws.com",
	}
	s3Client, err := minio.New(config)
	if err != nil {
		log.Fatalln(err)
	}
	for object := range s3Client.ListObjects("nothinglikeanything", "test", true) {
		if object.Err != nil {
			log.Fatalln(object.Err)
		}
		log.Println(object.Stat.Key)
	}
}
```

**Comment by harshavardhana** [comment]
It does look like S3 favors '.' over '/'

**Comment by harshavardhana** [comment]
Fixed in #1265
